fe/fe-core/src/main/java/com/starrocks/catalog/MaterializedView.java

```diff
@@ -1595,30 +1596,46 @@ public void gsonPreProcess() throws IOException {
     @Override
     public void gsonPostProcess() throws IOException {
         super.gsonPostProcess();
+        // only range partition need to recover partitionRefTableExprs
+        if (!(partitionInfo instanceof ExpressionRangePartitionInfo)) {
+            return;
+        }
+        ExpressionRangePartitionInfo expressionRangePartitionInfo = (ExpressionRangePartitionInfo) partitionInfo;
+        Type partitionColType = expressionRangePartitionInfo.getPartitionColumns().get(0).getType();
+
+        // for single ref base table, we need to recover partitionRefTableExprs
         partitionRefTableExprs = new ArrayList<>();
         if (serializedPartitionRefTableExprs != null) {
             for (GsonUtils.ExpressionSerializedObject expressionSql : serializedPartitionRefTableExprs) {
                 if (expressionSql != null) {
-                    partitionRefTableExprs.add(
-                            SqlParser.parseSqlToExpr(expressionSql.expressionSql, SqlModeHelper.MODE_DEFAULT));
+                    Expr partitionExpr = SqlParser.parseSqlToExpr(expressionSql.expressionSql, SqlModeHelper.MODE_DEFAULT);
+                    if (partitionExpr.getType() == Type.INVALID) {
+                        partitionExpr.setType(partitionColType);
+                    }
+                    partitionRefTableExprs.add(partitionExpr);
                 }
             }
         }
+
+        // for multi ref base tables, we need to recover partitionExprMaps
         partitionExprMaps = Maps.newHashMap();
         if (serializedPartitionExprMaps != null) {
             for (Map.Entry<GsonUtils.ExpressionSerializedObject, GsonUtils.ExpressionSerializedObject> entry :
                     serializedPartitionExprMaps.entrySet()) {
                 if (entry.getKey() != null && entry.getValue() != null) {
-                    partitionExprMaps.put(
-                            SqlParser.parseSqlToExpr(entry.getKey().expressionSql, SqlModeHelper.MODE_DEFAULT),
-                            (SlotRef) SqlParser.parseSqlToExpr(entry.getValue().expressionSql, SqlModeHelper.MODE_DEFAULT)
-                    );
+                    Expr partitionExpr = SqlParser.parseSqlToExpr(entry.getKey().expressionSql, SqlModeHelper.MODE_DEFAULT);
+                    if (partitionExpr.getType() == Type.INVALID) {
```

Comment:
why partitionExpr.getType is invalid, can we use a Analyzer to resolve these type?

Reply:
Like above, add more comments about this.

partitionRefTableExprs is not serialized for gson and only serializes the expression-sql instead which I think is more compatible for upgrade/downgrade.

so need to recover it in gsonPostProcess phase, expressionSql  is only parseSqlToExpr and no analyzed and needs to check expression's type here to make it more compatible for further usage.